modules/dicehub/endpoints/release.go

```diff
@@ -541,3 +814,287 @@ func getPermissionHeader(r *http.Request) (int64, error) {
+}
+
+// hasWriteAccess check whether user is project owner or project lead
+func (e *Endpoints) hasWriteAccess(identity apistructs.IdentityInfo, projectID int64, isProjectRelease bool, applicationID int64) (bool, error) {
+	req := &apistructs.ScopeRoleAccessRequest{
+		Scope: apistructs.Scope{
+			Type: apistructs.ProjectScope,
+			ID:   strconv.FormatInt(projectID, 10),
+		},
+	}
+	rsp, err := e.bdl.ScopeRoleAccess(identity.UserID, req)
+	if err != nil {
+		return false, err
+	}
+
+	hasProjectAccess := false
+	for _, role := range rsp.Roles {
+		if role == bundle.RoleProjectOwner || role == bundle.RoleProjectLead || role == bundle.RoleProjectPM {
+			hasProjectAccess = true
+			break
+		}
+	}
+
+	if isProjectRelease || hasProjectAccess {
+		return hasProjectAccess, nil
+	}
+
+	req = &apistructs.ScopeRoleAccessRequest{
+		Scope: apistructs.Scope{
+			Type: apistructs.AppScope,
+			ID:   strconv.FormatInt(applicationID, 10),
+		},
+	}
+	rsp, err = e.bdl.ScopeRoleAccess(identity.UserID, req)
+	if err != nil {
+		return false, err
+	}
+
+	hasAppAccess := false
+	for _, role := range rsp.Roles {
+		if role == bundle.RoleAppOwner || role == bundle.RoleAppLead {
+			hasAppAccess = true
+			break
+		}
+	}
+	return hasAppAccess, nil
+}
+
+func unmarshalApplicationReleaseList(str string) ([]string, error) {
+	var list []string
+	if err := json.Unmarshal([]byte(str), &list); err != nil {
+		return nil, err
+	}
+	return list, nil
+}
+
+func makeMetadata(release *dbclient.Release, appReleases []dbclient.Release) ([]byte, error) {
+	appList := make(map[string]apistructs.AppMetadata)
+	for i := range appReleases {
+		labels := make(map[string]string)
+		if err := json.Unmarshal([]byte(appReleases[i].Labels), &labels); err != nil {
+			return nil, errors.Errorf("failed to unmarshal labels for release %s", appReleases[i].ReleaseID)
+		}
+		appList[appReleases[i].ApplicationName] = apistructs.AppMetadata{
+			GitBranch:        labels["gitBranch"],
+			GitCommitID:      labels["gitCommitId"],
+			GitCommitMessage: labels["gitCommitMessage"],
+			GitRepo:          labels["gitRepo"],
+			ChangeLog:        appReleases[i].Markdown,
+		}
+	}
+	releaseMeta := apistructs.ReleaseMetadata{
+		Version:   release.Version,
+		Desc:      release.Desc,
+		ChangeLog: release.Markdown,
+		AppList:   appList,
+	}
+	return yaml.Marshal(releaseMeta)
+}
+
+func parseMetadata(file io.ReadCloser) (*apistructs.ReleaseMetadata, error) {
+	var metadata apistructs.ReleaseMetadata
+	found := false
+	reader := tar.NewReader(file)
+	for {
+		hdr, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		buf := bytes.Buffer{}
+		if _, err = io.Copy(&buf, reader); err != nil {
+			return nil, err
+		}
+
+		splits := strings.Split(hdr.Name, "/")
+		if len(splits) == 2 && splits[1] == "metadata.yml" {
+			if err := yaml.Unmarshal(buf.Bytes(), &metadata); err != nil {
+				return nil, err
+			}
+			found = true
+			break
+		}
+	}
+	if !found {
```

Comment:
When a code file exceeds 1000 lines, you should consider refactoring the code

Reply:
Start a new pull request to fix it